packages/flutter/lib/src/material/carousel.dart

```diff
@@ -0,0 +1,751 @@
+    super.didUpdateWidget(oldWidget);
+    if (widget.controller != oldWidget.controller) {
+      oldWidget.controller?._detach(this);
+      if (widget.controller != null) {
+        _internalController?._detach(this);
+        _internalController = null;
+        widget.controller?._attach(this);
+      } else { // widget.controller == null && oldWidget.controller != null
+        _internalController = CarouselController();
+        _controller._attach(this);
+      }
+    }
+    if (widget.itemExtent != oldWidget.itemExtent) {
+      _itemExtent = widget.itemExtent;
+    }
+  }
+
+  @override
+  void dispose() {
+    _controller._detach(this);
+    if (widget.controller == null) {
+      _controller.dispose();
+    }
+    super.dispose();
+  }
+
+  AxisDirection _getDirection(BuildContext context) {
+    switch (widget.scrollDirection) {
+      case Axis.horizontal:
+        assert(debugCheckHasDirectionality(context));
+        final TextDirection textDirection = Directionality.of(context);
+        final AxisDirection axisDirection = textDirectionToAxisDirection(textDirection);
+        return widget.reverse ? flipAxisDirection(axisDirection) : axisDirection;
+      case Axis.vertical:
+        return widget.reverse ? AxisDirection.up : AxisDirection.down;
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final ThemeData theme = Theme.of(context);
+    final AxisDirection axisDirection = _getDirection(context);
+    final ScrollPhysics physics = widget.itemSnapping
+      ? const CarouselScrollPhysics()
+      : ScrollConfiguration.of(context).getScrollPhysics(context);
+    final EdgeInsets effectivePadding = widget.padding ?? const EdgeInsets.all(4.0);
+    final Color effectiveBackgroundColor = widget.backgroundColor ?? Theme.of(context).colorScheme.surface;
+    final double effectiveElevation = widget.elevation ?? 0.0;
+    final ShapeBorder effectiveShape = widget.shape
+      ?? const RoundedRectangleBorder(
+        borderRadius: BorderRadius.all(Radius.circular(28.0))
+      );
+
+    return LayoutBuilder(
+      builder: (BuildContext context, BoxConstraints constraints) {
+        final double mainAxisExtent = switch (widget.scrollDirection) {
+          Axis.horizontal => constraints.maxWidth,
+          Axis.vertical => constraints.maxHeight,
+        };
+        _itemExtent = clampDouble(_itemExtent, 0, mainAxisExtent);
+
+        return Scrollable(
+          axisDirection: axisDirection,
+          controller: _controller,
+          physics: physics,
+          viewportBuilder: (BuildContext context, ViewportOffset position) {
+            return Viewport(
+              cacheExtent: 0.0,
+              cacheExtentStyle: CacheExtentStyle.viewport,
+              axisDirection: axisDirection,
+              offset: position,
+              clipBehavior: Clip.antiAlias,
+              slivers: <Widget>[
+                _SliverFixedExtentCarousel(
```

Comment:
Following up on our chat about ScrollView as a base class. Since this does not use a scroll view, the concept of primary (PrimaryScrollController) and keyboard dismiss behavior is omitted. Since Carousel defaults to a horizontal scroll direction, this may be less of an issue, but just one to consider for the vertical edge cases in the spec. Maybe we can just file and issue to follow up on if folks want these features? This combination of Scrollable+Viewport could probably be switched out for a CustomScrollView. 

I think the reason why PageView, which this current implementation is modeled after, does not utilize ScrollView as a base class is because PageView fills the entire viewport with its children. Or, one child of a PageView is displayed at a time. That and PageView primarily scrolls horizontally.

Reply:
Created a feature-request issue: https://github.com/flutter/flutter/issues/149765!